src/GitRepository.cc

```diff
@@ -0,0 +1,447 @@
+ * by the Free Software Foundation, either version 2.1 of the license, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+// SPDX-License-Identifier: LGPL-2.1-or-later
+
+#include <iostream>
+#include <vector>
+
+#include <git2.h>
+
+#include "gul14/cat.h"
+#include "taskolib/exceptions.h"
+#include "taskolib/GitRepository.h"
+
+
+namespace git {
+
+
+GitRepository::GitRepository(const std::filesystem::path& file_path)
+{
+    //init libgit library
+    git_libgit2_init();
+ 
+    //init member variables
+    repo_path_ = file_path;
+
+    // init or reload repository
+    init(file_path);
+}
+
+
+GitRepository::~GitRepository()
+{
+    repo_.reset();
+    my_signature_.reset();
+    git_libgit2_shutdown();
+}
+
+void GitRepository::reset_repo()
+{
+    repo_.reset();
+    my_signature_.reset();
+
+    // initialize repo and signature again
+    init(repo_path_);
+}
+
+std::string GitRepository::get_last_commit_message()
+{
+    const LibGitPointer<git_commit> commit = get_commit();
+    return std::string(git_commit_message(commit.get()));
+}
+
+std::filesystem::path GitRepository::get_path() const
+{
+    return repo_path_;
+}
+
+git_repository* GitRepository::get_repo()
+{
+    return repo_.get();
+}
+
+void GitRepository::update()
+{
+    LibGitPointer<git_index> index{repository_index(repo_.get())};
+
+    char *paths[] = {const_cast<char*>("*")};
+    git_strarray array = {paths, 1};
+
+    // update index to check for files
+    git_index_update_all(index.get(), &array, nullptr, nullptr);
+    git_index_write(index.get());
+}
+
+
+void GitRepository::init(const std::filesystem::path& file_path)
+{
+    repo_=repository_open(repo_path_);
+
+    // if repository does not exist
+    if (repo_.get() == nullptr) 
+    {
+        // create repository
+        //2nd argument: false so that .git folder is created in given path
+        repo_ = repository_init(file_path.c_str(), false);
+        if (repo_.get()==nullptr) throw task::Error("Git init failed.");
+
+        // create signature
+        my_signature_ = signature_default(repo_.get());
+        if (my_signature_.get()==nullptr) signature_new("Taskomat", "taskomat@desy.de", 123456789, 0);
+
+        // update files in directory
+        update();
+
+        // make initial commit
+        commit_initial();
+    }
+    // if repository already exists
+    else
+    {
+        // intialize the signature
+        my_signature_ = signature_default(repo_.get());
+        if (my_signature_.get() == nullptr) my_signature_ = signature_new("Taskomat", "taskomat@desy.de", 123456789, 0);
```

Comment:
... and since we do this dance twice, maybe this should go into a separate function.